### PR TITLE
feat: allow customizing browser data location

### PR DESCRIPTION
**Pull request by zcbenz**
#### Description of Change

This PR adds a new `sessionData` path to `app.setPath`, which allows changing the location where browser data like cookies and disk cache are stored.

This is very useful for apps that do not store user information in browser, as they treat browser data as cache and want to store them in cache locations, see also https://github.com/electron/electron/issues/1404 and https://github.com/electron/electron/issues/8124.

Before this PR apps can achieve similar effect by changing the path of `userData`, but we should separate `sessionData` from `userData`, as apps may want to store user information in `userData` while storing browser data in cache location.

This PR is modified from https://github.com/electron/electron/pull/30665.

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/main/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: Add `browserData` to `app.setPath/getPath`.

### CODE REVIEW

**Review by miniak — APPROVED** [review]

File: shell/app/electron_main_delegate.cc [hunk 1]
```diff
@@ -134,11 +134,18 @@ bool ElectronPathProvider(int key, base::FilePath* result) {
...
         return false;
       cur = cur.Append(base::FilePath::FromUTF8Unsafe("Dictionaries"));
       create_dir = true;
       break;
+    case DIR_BROWSER_DATA: {
+      // By default and for backward, equivalent to DIR_USER_DATA.
+      if (!base::PathService::Get(chrome::DIR_USER_DATA, &cur))
+        return false;
+      create_dir = true;
+      break;
```
Comment (nornagon) [inline shell/app/electron_main_delegate.cc:147]:
```suggestion
      return base::PathService::Get(chrome::DIR_USER_DATA, &cur);
```

I think `chrome::DIR_USER_DATA` already [creates the directory if it doesn't exist](https://source.chromium.org/chromium/chromium/src/+/main:chrome/common/chrome_paths.cc;l=165;drc=99f36136e05dbf96fa947c21feb467f60c357558), so we don't need to repeat the check.

File: shell/browser/browser.cc [hunk 2]
```diff
@@ -190,6 +190,9 @@ void Browser::DidFinishLaunching(base::DictionaryValue launch_info) {
   base::FilePath user_data;
   if (base::PathService::Get(chrome::DIR_USER_DATA, &user_data))
     base::CreateDirectoryAndGetError(user_data, nullptr);
+  base::FilePath browser_data;
+  if (base::PathService::Get(DIR_BROWSER_DATA, &browser_data))
+    base::CreateDirectoryAndGetError(browser_data, nullptr);
```
Comment (nornagon) [inline shell/browser/browser.cc:195]: Why is this needed? The directory should be created on demand, right?

File: docs/api/app.md [hunk 3]
```diff
@@ -637,6 +637,8 @@ Returns `string` - The current application directory.
     * `~/Library/Application Support` on macOS
   * `userData` The directory for storing your app's configuration files, which by
     default it is the `appData` directory appended with your app's name.
+  * `browserData` The directory for storing browsing data, such as localStorage,
+    cookies, and disk cache. By default this points to `userData`.
```
Comment (nornagon) [inline docs/api/app.md:641]: quite confused about this description, because it seems to me that localStorage and cookies should be in the "user data" directory.

**Review by nornagon — CHANGES_REQUESTED** [review]

File: docs/api/app.md (hunk 3 above)
Comment (zcbenz) [reply docs/api/app.md:641]: Chromium does not provide a way to set separate locations for localStorage and disk cache, so we currently put them all under `userData`. This PR means to provide a way to change the location of localStorage and disk cache, as there are apps willing to put them under cache locations (https://github.com/electron/electron/issues/1404).

In future I would like to add a new `browserCache` path to only set location for things like disk cache, but currently it is not practical as it requires a large patch to Chromium.

Comment (nornagon) [reply docs/api/app.md:641]: Then I'm not sure I see the point of this change then? What then is stored in `userData`?

Comment (zcbenz) [reply docs/api/app.md:641]: For example in VS Code, user configurations are stored as JSON under `userData`, while there is no user information stored in localStorage or cookies.

Also apart from browser data, things like crash reports and app logs are currently stored under `userData`, the `requestSingleInstance` API also relies on `userData` to exchange data.

For apps don't rely on localStorage or cookies to store user data, it is necessary to separate `browserData` from `userData`.

File: shell/browser/browser.cc (hunk 2 above)
Comment (zcbenz) [reply shell/browser/browser.cc:195]: If user sets a custom dir, it is not created on demand. There is code ensuring DIR_USER_DATA  to be created here for the same reason.

File: docs/api/app.md (hunk 3 above)
Comment (nornagon) [reply docs/api/app.md:641]: So it looks like this PR would group the following things under the new `browserData` directory:
- dictionaries
- browser prefs (`Local State`). not sure exactly what is stored in here, maybe things like devtools state?
- network service state. also not sure exactly what data is stored here. dns cache perhaps?
- ElectronBrowserContext's path, which is all the user profile data like localStorage, cookies, etc.
- some devtools-related temporary unix sockets

and things that remain in `userData` are:
- crash reports
- some caches?
- anything else?

Is this accurate? This is a weird grouping. I feel like `browserData` is the wrong name for this. At the very least this should be extensively documented, including exactly which data is stored in which directory.

Comment (zcbenz) [reply docs/api/app.md:641]: The idea behind the grouping is, users should be able to move all files generated by `BrowserContext` to a different directory. So dictionaries, browser prefs, devtools files, network caches belong to the same group with localStorage and cookies, as they are only generated when a `BrowserContext` is created.

I agree `browserData` is confusing, "browser" means lots of things in Electron. How do you think using `sessionData` as name instead? It should be clear that it is where `Session` code writes files to. If in future we add APIs to specify the path of `Session` to store data, it would point to `sessionData` by default.

  * `sessionData` The directory for storing data generated by `Session`, such as localStorage, cookies, disk cache, downloaded dictionaries, network state, devtools files. By default this points to `userData`.

Comment (nornagon) [reply docs/api/app.md:641]: I don't think downloaded dictionaries are specific to a Session?

Comment (nornagon) [reply docs/api/app.md:641]: Also, are we sure that no other "session-related" data is stored in userData? e.g. v8 code caches, IndexedDB data, service worker registration, etc. etc.

Comment (zcbenz) [reply docs/api/app.md:641]: Dictionaries are downloaded by Chromium automatically via the `ses.setSpellCheckerDictionaryDownloadURL` API, and all dictionary APIs belong to Session. In Chromium's context, the SpellcheckServiceFactory is attached to BrowserContext.

For other data, I can write tests to verify they are in the expected location.

**Comment by zcbenz** [comment]
I have renamed `browserData` to `sessionData`, and added tests to verify the locations of service worker, localStorage, etc. are the in the right location.

File: docs/api/app.md (hunk 3 above)
Comment (nornagon) [reply docs/api/app.md]: I'm somewhat worried about new session-related data being introduced in Chromium with no easy way for us to redirect where it's stored without patching Chromium. It also seems really easy to miss session-related data when implementing this. How can we be sure we caught all the cases?

Conceptually, though, the idea of a `sessionData` path doesn't seem totally unreasonable. It does raise the question from an API perspective of whether it should be possible to set session paths on a per-session basis.

File: docs/api/app.md [hunk 4]
```diff
@@ -637,6 +637,9 @@ Returns `string` - The current application directory.
     * `~/Library/Application Support` on macOS
   * `userData` The directory for storing your app's configuration files, which by
     default it is the `appData` directory appended with your app's name.
```
Comment (nornagon) [inline docs/api/app.md:639]: Let's expand this comment to describe how it differs from `sessionData`, since "user" data is now quite confusing.

**Review by jkleinsc — COMMENTED** [review]
API LGTM

File: docs/api/app.md [hunk 5]
```diff
@@ -635,8 +635,18 @@ Returns `string` - The current application directory.
     * `%APPDATA%` on Windows
     * `$XDG_CONFIG_HOME` or `~/.config` on Linux
     * `~/Library/Application Support` on macOS
-  * `userData` The directory for storing your app's configuration files, which by
-    default it is the `appData` directory appended with your app's name.
+  * `userData` The directory for storing your app's configuration files, which
+    by default is the `appData` directory appended with your app's name. By
+    convention files storing user data should be written to this directory, and
+    it is not recommended to write large files here because some environments
+    may backup this directory to cloud storage.
```
Comment (nornagon) [inline docs/api/app.md:642]: Does Chromium have this problem? I would be surprised if Chrome's caches are uploaded to cloud storage on any platform...

Comment (nornagon) [reply docs/api/app.md:642]: Hang on—so Chrome can choose the session data directory per-profile, then! Chromium will put things under `userData/Profile 1`, `userData/Profile 2` and so on. Shouldn't we offer the same capability?

Comment (nornagon) [reply docs/api/app.md:642]: Hm, Chrome writes to `User Data/$partitionName`, right? `Default` is a name of a partition.

Comment (nornagon) [reply docs/api/app.md:642]: ahhhh, right, got it. Thanks, that makes sense!

File: docs/api/app.md [hunk 6]
```diff
@@ -635,8 +635,18 @@ Returns `string` - The current application directory.
...
+  * `sessionData` The directory for storing data generated by `Session`, such
+    as localStorage, cookies, disk cache, downloaded dictionaries, network
+    state, devtools files. By default this points to `userData`. Chromium may
+    write very large disk cache here, so if your app does not rely on browser
+    storage like localStorage or cookies to save user data, it is recommended
+    to set this directory to other locations to avoid polluting the `userData`
```
Comment (nornagon) [inline docs/api/app.md:648]: > other locations

I think it would be a good idea to recommend a particular location here!

**Review by nornagon — APPROVED** [review]

**Comment by jkleinsc** [comment]
Merging as API review has been approved but is still showing as pending for some reason.
